File: Window/windows-native.cc
#include <windows.h>
#include <napi.h>
#include <thread>
#include <chrono>

namespace {

/**
 * Convert screen coordinates to POINT
 * Uses global screen coordinates with origin at top-left
 */
POINT ToPoint(double x, double y) {
  POINT p;
  p.x = static_cast<LONG>(x);
  p.y = static_cast<LONG>(y);
  return p;
}

bool gLeftDown = false;

/**
 * Move cursor to target position
 */
void MoveCursor(POINT point) {
  SetCursorPos(point.x, point.y);
}

POINT CurrentCursorPos() {
  POINT p;
  GetCursorPos(&p);
  return p;
}

/**
 * Send mouse input event
 */
void SendMouseInput(DWORD flags, LONG dx = 0, LONG dy = 0) {
  INPUT input = {0};
  input.type = INPUT_MOUSE;
  input.mi.dwFlags = flags;
  input.mi.dx = dx;
  input.mi.dy = dy;
  input.mi.mouseData = 0;
  input.mi.dwExtraInfo = 0;
  input.mi.time = 0;
  SendInput(1, &input, sizeof(INPUT));
}

/**
 * Send mouse event at absolute coordinates
 */
void SendMouseEventAbsolute(DWORD flags, double x, double y) {
  // Convert to normalized absolute coordinates (0-65535)
  int screenWidth = GetSystemMetrics(SM_CXSCREEN);
  int screenHeight = GetSystemMetrics(SM_CYSCREEN);
  
  LONG normalizedX = static_cast<LONG>((x * 65535) / screenWidth);
  LONG normalizedY = static_cast<LONG>((y * 65535) / screenHeight);
  
  INPUT input = {0};
  input.type = INPUT_MOUSE;
  input.mi.dwFlags = flags | MOUSEEVENTF_ABSOLUTE;
  input.mi.dx = normalizedX;
  input.mi.dy = normalizedY;
  input.mi.mouseData = 0;
  input.mi.dwExtraInfo = 0;
  input.mi.time = 0;
  SendInput(1, &input, sizeof(INPUT));
  
  std::this_thread::sleep_for(std::chrono::milliseconds(10));
}

/**
 * Send drag path from start to end position
 * Segmented dragging to avoid single position jump being ignored by some applications
 */
void SendDragPath(POINT from, POINT to) {
  int steps = 8;
  for (int i = 1; i <= steps; ++i) {
    double t = static_cast<double>(i) / static_cast<double>(steps);
    LONG x = static_cast<LONG>(from.x + (to.x - from.x) * t);
    LONG y = static_cast<LONG>(from.y + (to.y - from.y) * t);
    
    SetCursorPos(x, y);
    SendMouseInput(MOUSEEVENTF_MOVE);
    std::this_thread::sleep_for(std::chrono::milliseconds(5));
  }
}

Napi::Value MoveWrapped(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  if (info.Length() < 2 || !info[0].IsNumber() || !info[1].IsNumber()) {
    Napi::TypeError::New(env, "move(x, y) expects two numbers").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  double x = info[0].As<Napi::Number>().DoubleValue();
  double y = info[1].As<Napi::Number>().DoubleValue();
  POINT p = ToPoint(x, y);
  
  // Check if left button is currently down
  bool isLeftDown = gLeftDown || (GetAsyncKeyState(VK_LBUTTON) & 0x8000);
  
  if (isLeftDown) {
    POINT cur = CurrentCursorPos();
    SendDragPath(cur, p);
  } else {
    MoveCursor(p);
    SendMouseInput(MOUSEEVENTF_MOVE);
  }
  return env.Undefined();
}

/**
 * NAPI wrapper for mouse down event
 * If coordinates are provided, use them; otherwise use current mouse position
 */
Napi::Value MouseDownWrapped(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  if (info.Length() >= 2 && info[0].IsNumber() && info[1].IsNumber()) {
    double x = info[0].As<Napi::Number>().DoubleValue();
    double y = info[1].As<Napi::Number>().DoubleValue();
    MoveCursor(ToPoint(x, y));
  }
  
  gLeftDown = true;
  SendMouseInput(MOUSEEVENTF_LEFTDOWN);
  
  return env.Undefined();
}

/**
 * NAPI wrapper for mouse up event
 * If coordinates are provided, use them; otherwise use current mouse position
 */
Napi::Value MouseUpWrapped(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  if (info.Length() >= 2 && info[0].IsNumber() && info[1].IsNumber()) {
    double x = info[0].As<Napi::Number>().DoubleValue();
    double y = info[1].As<Napi::Number>().DoubleValue();
    MoveCursor(ToPoint(x, y));
  }
  
  gLeftDown = false;
  SendMouseInput(MOUSEEVENTF_LEFTUP);
  
  return env.Undefined();
}

Napi::Value ClickWrapped(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  if (info.Length() < 2 || !info[0].IsNumber() || !info[1].IsNumber()) {
    Napi::TypeError::New(env, "click(x, y) expects two numbers").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  double x = info[0].As<Napi::Number>().DoubleValue();
  double y = info[1].As<Napi::Number>().DoubleValue();
  POINT p = ToPoint(x, y);
  
  MoveCursor(p);
  std::this_thread::sleep_for(std::chrono::milliseconds(10));
  
  // Send down and up events with delays to simulate real click
  SendMouseInput(MOUSEEVENTF_LEFTDOWN);
  std::this_thread::sleep_for(std::chrono::milliseconds(20));
  SendMouseInput(MOUSEEVENTF_LEFTUP);
  
  return env.Undefined();
}

/**
 * Get current mouse position
 */
Napi::Value GetMousePosWrapped(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  POINT p = CurrentCursorPos();
  
  Napi::Object result = Napi::Object::New(env);
  result.Set("x", Napi::Number::New(env, p.x));
  result.Set("y", Napi::Number::New(env, p.y));
  
  return result;
}

/**
 * Get screen dimensions
 */
Napi::Value GetScreenSizeWrapped(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  int width = GetSystemMetrics(SM_CXSCREEN);
  int height = GetSystemMetrics(SM_CYSCREEN);
  
  Napi::Object result = Napi::Object::New(env);
  result.Set("width", Napi::Number::New(env, width));
  result.Set("height", Napi::Number::New(env, height));
  
  return result;
}

/**
 * Overlay window management using Windows API
 */
HWND gOverlayWindow = NULL;
HBRUSH gBorderBrush = NULL;

LRESULT CALLBACK OverlayWndProc(HWND hwnd, UINT msg, WPARAM wParam, LPARAM lParam) {
  switch (msg) {
    case WM_PAINT: {
      PAINTSTRUCT ps;
      HDC hdc = BeginPaint(hwnd, &ps);
      
      // Get window dimensions
      RECT rect;
      GetClientRect(hwnd, &rect);
      
      // Create purple pen for border
      HPEN hPen = CreatePen(PS_SOLID, 5, RGB(143, 83, 201)); // #8F53C9
      HPEN hOldPen = (HPEN)SelectObject(hdc, hPen);
      
      // Set transparent background
      SelectObject(hdc, GetStockObject(NULL_BRUSH));
      
      // Draw rounded rectangle border
      int radius = 20;
      RoundRect(hdc, 0, 0, rect.right, rect.bottom, radius, radius);
      
      SelectObject(hdc, hOldPen);
      DeleteObject(hPen);
      
      EndPaint(hwnd, &ps);
      return 0;
    }
    case WM_DESTROY:
      PostQuitMessage(0);
      return 0;
    default:
      return DefWindowProc(hwnd, msg, wParam, lParam);
  }
}

/**
 * Create overlay window with rounded border
 * Runs in a separate thread to avoid blocking
 */
/**
 * Create overlay window with rounded border
 * Runs in a separate thread to avoid blocking
 */
void CreateOverlayWindowThread() {
  HINSTANCE hInstance = GetModuleHandle(NULL);
  
  // Register window class
  WNDCLASSEXW wc = {0};
  wc.cbSize = sizeof(WNDCLASSEXW);
  wc.lpfnWndProc = OverlayWndProc;
  wc.hInstance = hInstance;
  wc.lpszClassName = L"OverlayWindowClass";
  wc.hbrBackground = (HBRUSH)GetStockObject(NULL_BRUSH);
  
  if (!RegisterClassExW(&wc)) {
    DWORD error = GetLastError();
    if (error != ERROR_CLASS_ALREADY_EXISTS) {
      return;
    }
  }
  
  // Get screen dimensions
  int screenWidth = GetSystemMetrics(SM_CXSCREEN);
  int screenHeight = GetSystemMetrics(SM_CYSCREEN);
  
  // Create layered window
  gOverlayWindow = CreateWindowExW(
    WS_EX_LAYERED | WS_EX_TRANSPARENT | WS_EX_TOPMOST | WS_EX_TOOLWINDOW,
    L"OverlayWindowClass",
    L"Overlay",
    WS_POPUP,
    0, 0, screenWidth, screenHeight,
    NULL, NULL, hInstance, NULL
  );
  
  if (!gOverlayWindow) {
    return;
  }
  
  // Make window transparent and click-through
  SetLayeredWindowAttributes(gOverlayWindow, RGB(0, 0, 0), 0, LWA_COLORKEY);
  SetLayeredWindowAttributes(gOverlayWindow, 0, 230, LWA_ALPHA);
  
  ShowWindow(gOverlayWindow, SW_SHOW);
  UpdateWindow(gOverlayWindow);
  
  // Message loop (non-blocking, just process existing messages)
  MSG msg;
  while (PeekMessage(&msg, NULL, 0, 0, PM_REMOVE)) {
    TranslateMessage(&msg);
    DispatchMessage(&msg);
  }
}

/**
 * NAPI wrapper for creating overlay window
 */
Napi::Value CreateOverlayWrapped(const Napi::CallbackInfo& info) {
  if (gOverlayWindow) {
    ShowWindow(gOverlayWindow, SW_SHOW);
    SetWindowPos(gOverlayWindow, HWND_TOPMOST, 0, 0, 0, 0, 
                 SWP_NOMOVE | SWP_NOSIZE | SWP_SHOWWINDOW);
  } else {
    std::thread(CreateOverlayWindowThread).detach();
  }
  return info.Env().Undefined();
}

Napi::Object Init(Napi::Env env, Napi::Object exports) {
  exports.Set("click", Napi::Function::New(env, ClickWrapped));
  exports.Set("move", Napi::Function::New(env, MoveWrapped));
  exports.Set("mouseDown", Napi::Function::New(env, MouseDownWrapped));
  exports.Set("mouseUp", Napi::Function::New(env, MouseUpWrapped));
  exports.Set("getMousePos", Napi::Function::New(env, GetMousePosWrapped));
  exports.Set("getScreenSize", Napi::Function::New(env, GetScreenSizeWrapped));
  exports.Set("createOverlayFrame", Napi::Function::New(env, CreateOverlayWrapped));
  return exports;
}

}  // namespace

NODE_API_MODULE(NODE_GYP_MODULE_NAME, Init)

File: windows-native.js
const ffi = require('ffi-napi');
const ref = require('ref-napi');

// Define Windows API types
const DWORD = ref.types.uint32;
const LONG = ref.types.long;
const WORD = ref.types.uint16;
const UINT = ref.types.uint32;

// INPUT structure types
const INPUT_MOUSE = 0;
const INPUT_KEYBOARD = 1;

// Mouse event flags
const MOUSEEVENTF_MOVE = 0x0001;
const MOUSEEVENTF_LEFTDOWN = 0x0002;
const MOUSEEVENTF_LEFTUP = 0x0004;
const MOUSEEVENTF_RIGHTDOWN = 0x0008;
const MOUSEEVENTF_RIGHTUP = 0x0010;
const MOUSEEVENTF_MIDDLEDOWN = 0x0020;
const MOUSEEVENTF_MIDDLEUP = 0x0040;
const MOUSEEVENTF_ABSOLUTE = 0x8000;

// Define MOUSEINPUT structure
const MOUSEINPUT = ref.types.void;

// Define INPUT structure
const INPUT = ref.types.void;

// Load User32.dll
const user32 = ffi.Library('user32', {
    'SetCursorPos': ['bool', ['int', 'int']],
    'GetCursorPos': ['bool', [ref.refType(ref.types.void)]],
    'SendInput': ['uint', ['uint', ref.refType(INPUT), 'int']],
    'GetSystemMetrics': ['int', ['int']]
});

// System metrics constants
const SM_CXSCREEN = 0;
const SM_CYSCREEN = 1;

/**
 * Get screen dimensions
 */
function getScreenSize() {
    const width = user32.GetSystemMetrics(SM_CXSCREEN);
    const height = user32.GetSystemMetrics(SM_CYSCREEN);
    return { width, height };
}

/**
 * Convert absolute coordinates to normalized coordinates (0-65535)
 */
function toNormalizedCoords(x, y) {
    const screen = getScreenSize();
    const normalizedX = Math.round((x * 65535) / screen.width);
    const normalizedY = Math.round((y * 65535) / screen.height);
    return { x: normalizedX, y: normalizedY };
}

/**
 * Simulates moving the mouse to the specified (x, y) coordinates.
 * Uses SetCursorPos for direct, immediate positioning.
 *
 * @param {number} x - The x-coordinate to move to.
 * @param {number} y - The y-coordinate to move to.
 */
function move(x, y) {
    try {
        const result = user32.SetCursorPos(Math.round(x), Math.round(y));
        if (!result) {
            console.error('SetCursorPos failed');
        }
    } catch (error) {
        console.error('Error moving mouse:', error);
    }
}

/**
 * Simulates a virtual click at the specified (x, y) coordinates.
 *
 * @param {number} x - The x-coordinate of the click.
 * @param {number} y - The y-coordinate of the click.
 */
function click(x, y) {
    try {
        // Move to position
        move(x, y);
        
        // Small delay to ensure position is set
        const start = Date.now();
        while (Date.now() - start < 10) {}
        
        // Perform click (down then up)
        mouseDown();
        
        // Small delay between down and up
        const start2 = Date.now();
        while (Date.now() - start2 < 50) {}
        
        mouseUp();
    } catch (error) {
        console.error('Error clicking mouse:', error);
    }
}

/**
 * Simulates a mouse down event (press left button).
 */
function mouseDown() {
    try {
        // Create INPUT structure for mouse down
        const inputSize = 28; // Size of INPUT structure
        const input = Buffer.alloc(inputSize);
        
        // Set type to INPUT_MOUSE
        input.writeUInt32LE(INPUT_MOUSE, 0);
        
        // Set dwFlags to MOUSEEVENTF_LEFTDOWN
        input.writeUInt32LE(MOUSEEVENTF_LEFTDOWN, 8);
        
        const result = user32.SendInput(1, input, inputSize);
        if (result === 0) {
            console.error('SendInput (mouseDown) failed');
        }
    } catch (error) {
        console.error('Error in mouseDown:', error);
    }
}

/**
 * Simulates a mouse up event (release left button).
 */
function mouseUp() {
    try {
        // Create INPUT structure for mouse up
        const inputSize = 28; // Size of INPUT structure
        const input = Buffer.alloc(inputSize);
        
        // Set type to INPUT_MOUSE
        input.writeUInt32LE(INPUT_MOUSE, 0);
        
        // Set dwFlags to MOUSEEVENTF_LEFTUP
        input.writeUInt32LE(MOUSEEVENTF_LEFTUP, 8);
        
        const result = user32.SendInput(1, input, inputSize);
        if (result === 0) {
            console.error('SendInput (mouseUp) failed');
        }
    } catch (error) {
        console.error('Error in mouseUp:', error);
    }
}

/**
 * Get current mouse position
 */
function getMousePos() {
    try {
        const point = Buffer.alloc(8); // POINT structure is 8 bytes (2 LONGs)
        const result = user32.GetCursorPos(point);
        
        if (result) {
            const x = point.readInt32LE(0);
            const y = point.readInt32LE(4);
            return { x, y };
        } else {
            console.error('GetCursorPos failed');
            return { x: 0, y: 0 };
        }
    } catch (error) {
        console.error('Error getting mouse position:', error);
        return { x: 0, y: 0 };
    }
}

/**
 * Creates a border around the viewport to show what the user 
 * is currently sharing (Windows implementation).
 */
function createOverlayFrame() {
    const { spawn } = require('child_process');
    const path = require('path');
    
    const overlayPath = path.join(__dirname, 'Examples', 'example-overlay-windows.js');
    const electronPath = require('electron');
    
    try {
        spawn(electronPath, [overlayPath], {
            detached: true,
            stdio: 'ignore'
        }).unref();
    } catch (error) {
        console.error('Error creating overlay frame:', error);
    }
}

module.exports = { 
    click, 
    mouseDown, 
    mouseUp, 
    move, 
    getMousePos,
    createOverlayFrame,
    getScreenSize
};
